src/coprocessor/dag/storage/scanner.rs

```diff
@@ -0,0 +1,567 @@
+    working_range_begin_key: Vec<u8>,
+    working_range_end_key: Vec<u8>,
+}
+
+impl<T: Storage> RangesScanner<T> {
+    pub fn new(
+        storage: T,
+        ranges: Vec<Range>,
+        scan_backward_in_range: bool, // TODO: This can be const generics
+        is_key_only: bool,            // TODO: This can be const generics
+        is_scanned_range_aware: bool, // TODO: This can be const generics
+    ) -> RangesScanner<T> {
+        let ranges_len = ranges.len();
+        let ranges_iter = RangesIterator::new(ranges);
+        RangesScanner {
+            storage,
+            ranges_iter,
+            scan_backward_in_range,
+            is_key_only,
+            scanned_rows_per_range: Vec::with_capacity(ranges_len),
+            is_scanned_range_aware,
+            current_range_lower_key: Vec::with_capacity(KEY_BUFFER_CAPACITY),
+            current_range_upper_key: Vec::with_capacity(KEY_BUFFER_CAPACITY),
+            working_range_begin_key: Vec::with_capacity(KEY_BUFFER_CAPACITY),
+            working_range_end_key: Vec::with_capacity(KEY_BUFFER_CAPACITY),
+        }
+    }
+
+    /// Fetches next row.
+    // Note: This is not implemented over `Iterator` since it can fail.
+    // TODO: Change to use reference to avoid alloation and copy.
+    pub fn next(&mut self) -> Result<Option<OwnedKvPair>, Error> {
+        loop {
+            let range = self.ranges_iter.next();
+            let some_row = match range {
+                IterStatus::NewRange(Range::Point(r)) => {
+                    if self.is_scanned_range_aware {
+                        self.update_scanned_range_from_new_point(&r);
+                    }
+                    self.ranges_iter.notify_drained();
+                    self.scanned_rows_per_range.push(0);
+                    self.storage.get(self.is_key_only, r)?
+                }
+                IterStatus::NewRange(Range::Interval(r)) => {
+                    if self.is_scanned_range_aware {
+                        self.update_scanned_range_from_new_range(&r);
+                    }
+                    self.scanned_rows_per_range.push(0);
+                    self.storage
+                        .begin_scan(self.scan_backward_in_range, self.is_key_only, r)?;
+                    self.storage.scan_next()?
+                }
+                IterStatus::Continue => self.storage.scan_next()?,
+                IterStatus::Drained => {
+                    if self.is_scanned_range_aware {
+                        self.update_working_range_end_key();
+                    }
+                    return Ok(None); // drained
+                }
+            };
+            if self.is_scanned_range_aware {
+                self.update_scanned_range_from_scanned_row(&some_row);
+            }
+            if some_row.is_some() {
+                // Retrieved one row from point range or interval range.
+                if let Some(r) = self.scanned_rows_per_range.last_mut() {
```

Comment:
It's better to move `scanned_rows_per_range` related logic to `RangesIterator`.

Reply:
RangesIterator should only iterate ranges